### PR TITLE
Make duplicate index exceptions more descriptive

### PR DESCRIPTION
**Pull request by coroa**
# Please confirm that this PR has done the following:

- ~Tests Added~ -> no new behavior
- ~Documentation Added~ -> no changes
- ~Name of contributors Added to AUTHORS.rst~
- ~Description in RELEASE_NOTES.md Added~ -> I'd say not noteworthy enough! objections?

# Description of PR

`IamDataFrame.append` and (by transitivity) `pyam.concat` defaulted to using `pd.Series.append`'s `verify_integrity`, which lists only the first 3 levels of the overlapping multi-index.

This PR re-uses the index summary part of the `info` representation to give a high-level overview of the conflicts.

## Synthetic example

```python
df.append(df.filter(variable="Primary_Energy|Biomass", region=["DEU", "USA"]))
```

### Previous behaviour
```
[ ... 8-level stacktrace ... ]
ValueError: Indexes have overlapping values: MultiIndex([(       'PAC',             'PAC', 'DEU', ...),
            (       'PAC',             'PAC', 'DEU', ...),
            (       'PAC',             'PAC', 'DEU', ...),
            (       'PAC',             'PAC', 'DEU', ...),
            (       'PAC',             'PAC', 'DEU', ...),
            (       'PAC',             'PAC', 'DEU', ...),
            (       'PAC',             'PAC', 'DEU', ...),
            (       'PAC',             'PAC', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...),
            ('REMIND_1.7', 'CEMICS-1.5-CDR8', 'DEU', ...)],
           names=['model', 'scenario', 'region', 'variable', 'unit', 'year'])
```

## New behaviour
```
pyam.utils - ERROR: Timeseries data have overlapping values:
   model          scenario region                variable     unit  year
0  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2017
1  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2020
2  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2025
3  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2030
4  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2035
```

### CODE REVIEW

File: pyam/core.py [hunk 1]
```diff
@@ -437,7 +437,7 @@ def variables(self, include_units=False):
         )
 
     def append(self, other, ignore_meta_conflict=False, inplace=False,
-               **kwargs):
+               verify_integrity=True, **kwargs):
```
Comment (danielhuppmann) [inline pyam/core.py:440]: I would prefer to not (even optionally) allow the possibility to have duplicate rows  in the timeseries data...

File: pyam/core.py [hunk 2]
```diff
@@ -485,7 +487,9 @@ def append(self, other, ignore_meta_conflict=False, inplace=False,
         ret.meta = merge_meta(ret.meta, other.meta, ignore_meta_conflict)
 
         # append other.data (verify integrity for no duplicates)
-        _data = ret._data.append(other._data, verify_integrity=True)
+        _data = ret._data.append(other._data)
```
Comment (danielhuppmann) [inline pyam/core.py:490]: I guess you have to be careful here to not change `df._data` (if operating inplace) and then raising the error afterwards

**Review by danielhuppmann — CHANGES_REQUESTED** [review]
Thanks @coroa, I really like the idea to improve the error message! But take a look at the function, this is an alternative to have a more meaningful error message.

https://github.com/IAMconsortium/pyam/blob/9218247550916bb009c24655c6ffeeb6d7faa2c7/pyam/utils.py#L322

File: pyam/core.py (hunk 1 above)
Comment (coroa) [reply pyam/core.py:440]: `is_unique` on a long multiindex is actually quite an expensive operation. i would like to give the knowledgable user a safety belt off option where he is responsible for ensuring integrity or where he is responsible for checking integrity.

The main use case in my mind was a small improvement for pyam.concat where we call append with `verify_integrity = False` and then call `verify_index_integrity` once the final dataframe has been constructed. (Converting that function directly to `pd.concat` as proposed in #501 is way better though).

Comment (danielhuppmann) [reply pyam/core.py:440]: Fair enough... Doesn't stand in the way of doing both this improvement and improving `concat()` in the separate PR

File: pyam/core.py (hunk 2 above)
Comment (coroa) [reply pyam/core.py:490]: No you don't.

Pandas is very clear on when it operates in place and when it does not. The only inplace operations are:
1. __setitem__ (`df.loc[..] = ...`, `df[...] = ...`, ...) 
2. __setattr__  (`df.index = ...`, `df.name = ...`)
3. methods where you add an explicit `inplace=True` argument.

All other methods operate on *shallow* copies and there are non-inplace replacements for all of them, ie instead
of `df["test"] = "bla"`, you can use `df.assign(test="bla")`, instead of `df.index = ind`, you can use `df.set_axis(ind)`, and so on and not operate inplace at all. If one follows this behaviour systematically, one can actually avoid most of the deep copies, pyam is currently relying on.

Long story short, here we are not updating `ret._data` only `_data`.

Comment (danielhuppmann) [reply pyam/core.py:490]: Ok, got it, thanks for the clarification!

**Comment by coroa** [comment]
> this is an alternative to have a more meaningful error message.

If we make use of  `_raise_data_error("Indexes have overlapping values", overlap.to_frame(index=False))`, we get in the above example:

## Alternative behaviour with `_raise_data_error`

```
pyam.utils - ERROR: Indexes have overlapping values:
   model          scenario region                variable     unit  year
0  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2017
1  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2020
2  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2025
3  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2030
4  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2035
...

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)

[... 3 - level stacktrace ...]

ValueError: Indexes have overlapping values:
   model          scenario region                variable     unit  year
0  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2017
1  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2020
2  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2025
3  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2030
4  IMAGE  SSP1-19-SPA1-V17    DEU  Primary_Energy|Biomass  EJ / yr  2035
...

Which gives you less of an overview, in my taste, but might bring about the error better?

**Comment by danielhuppmann** [comment]
> but might bring about the error better?

I agree that this is closer to the kind of format for which a user has an intuition...

One suggestion: I would write "Timeseries data ..." instead of "Indexes ..." - most users probably don't even know that there is an indexed pd.Series in the background.

**Comment by coroa** [comment]
Grr ... CI fails due to connection problems with ixmp. Probably wait and restart. Do you have a re-run job button somewhere or do i have to push a fake commit?

**Comment by danielhuppmann** [comment]
Triggered CI again and gave you "write" permissions to the repo - that should do the trick that you can do it next time...

(I added exceptions for the World-Bank-API, not sure if we want to also add those for the IIASA Scenario Explorer test instance).

File: pyam/core.py [hunk 3]
```diff
@@ -477,6 +484,8 @@ def append(self, other, ignore_meta_conflict=False, inplace=False, **kwargs):
             any meta columns present in `self` and `other` are not identical.
         inplace : bool, default False
             If True, do operation inplace and return None
+        verify_integrity : bool, default True
```
Comment (danielhuppmann) [inline pyam/core.py:487]: Per numpydoc convention, use optional instead of the value (which is in the signature anyway, see section 4 at https://numpydoc.readthedocs.io/en/latest/format.html

Can you update this for all kwargs `append()` function? (The docstrings are currently not consistent, so I try to harmonize it towards numpydoc whenever a function is touched).

**Review by danielhuppmann — APPROVED** [review]
Looks good, thanks @coroa!

**Comment by danielhuppmann** [comment]
I also edited the description of the PR to mimic the new behavior - @coroa, feel free to merge!
